tests/src/Unit/CollaboraDiscoveryTest.php

```diff
@@ -28,12 +29,167 @@ public function testWopiClientUrl(): void {
+    $file = dirname(__DIR__, 2) . '/fixtures/discovery.xml';
+    $xml = file_get_contents($file);
+    $this->assertSame(98, preg_match_all('@<app( [^>]+)* name="([^"]+)"@', $xml, $matches));
+    $this->assertSame('application/vnd.ms-excel', $matches[2][9]);
+    $mimetypes = array_unique($matches[2]);
+    $mimetypes = array_diff($mimetypes, ['Capabilities']);
+    sort($mimetypes);
+    $discovery = $this->getDiscoveryFromXml($xml);
+    $known_url = 'http://collabora.test:9980/browser/61cf2b4/cool.html?';
+    $supported_action_types = [];
+    foreach ($mimetypes as $mimetype) {
+      $type_supported_actions = [];
+      foreach (['edit', 'view_comment', 'view'] as $action) {
+        $url = $discovery->getWopiClientURL($mimetype, $action);
+        if ($url !== NULL) {
+          $this->assertSame($known_url, $url);
+          $type_supported_actions[] = $action;
+        }
+      }
+      sort($type_supported_actions);
+      $supported_action_types[implode(',', $type_supported_actions)][] = $mimetype;
+    }
+    ksort($supported_action_types);
+    $this->assertSame([
+      '' => [
+        'application/vnd.oasis.opendocument.formula',
+        'application/vnd.sun.xml.math',
+        'math',
+      ],
+      'edit' => [
+        'application/msword',
+        'application/vnd.ms-excel',
+        'application/vnd.ms-excel.sheet.binary.macroEnabled.12',
+        'application/vnd.ms-excel.sheet.macroEnabled.12',
+        'application/vnd.ms-powerpoint',
+        'application/vnd.ms-powerpoint.presentation.macroEnabled.12',
+        'application/vnd.ms-word.document.macroEnabled.12',
+        'application/vnd.oasis.opendocument.chart',
+        'application/vnd.oasis.opendocument.graphics',
+        'application/vnd.oasis.opendocument.graphics-flat-xml',
+        'application/vnd.oasis.opendocument.presentation',
+        'application/vnd.oasis.opendocument.presentation-flat-xml',
+        'application/vnd.oasis.opendocument.spreadsheet',
+        'application/vnd.oasis.opendocument.spreadsheet-flat-xml',
+        'application/vnd.oasis.opendocument.text',
+        'application/vnd.oasis.opendocument.text-flat-xml',
+        'application/vnd.oasis.opendocument.text-master',
+        'application/vnd.oasis.opendocument.text-web',
+        'application/vnd.openxmlformats-officedocument.presentationml.presentation',
+        'application/vnd.openxmlformats-officedocument.presentationml.slideshow',
+        'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
+        'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
+        'application/vnd.sun.xml.draw',
+        'application/x-dbase',
+        'application/x-dif-document',
+        'text/csv',
+        'text/plain',
+        'text/rtf',
+        'text/spreadsheet',
+        'writer-web',
+      ],
+      'edit,view' => [
+        'calc',
+        'impress',
+        'writer',
+        'writer-global',
```

Comment:
These ones are not really MIME types, and they tend to have multiple `<action>` tags with `ext=".."` attribute.
In the default discovery.xml from Collabora they all have the same url, but this could be customized.
To make that work we would have to also pass extension name. But if the MIME type is never 'writer' or 'impress' etc for a given file entity, then this part is rather pointless.

Reply:
Not sure I understand what's this about.

Reply:
So, the array has the following structure:
`$arr[implode(',', $supported_actions)][] = $mime_type;`
E.g.
`$arr['view_comment'][] = 'application/pdf';` means that for application/pdf we have _only_ view_comment action.

The MIME types are taken from 'name' attributes in the `<app name="...">` tags.

But, some of the 'name' attributes in `<app name="...">` are not really MIME types. E.g. `<app name="writer">`. You can expand the xml above to see.